### PR TITLE
feat(unlock-app) #8362 - add custom recipient

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description
This PR is made to add the custom recipient on purchase. (#8362)
- Added “Advanced” button that shows Recipient input where users can put ENS domain or wallet address 
    - If `advance mode` is active and the recipient is not valid, buttons will be disabled with an error message
    - if `advance mode` is active and the recipient is valid, buttons will be enabled, and a message will show that the recipient is valid 
    - if `advanced mode` is active and the transaction is pending, the input for the custom address will be disabled

@searchableguy any suggestion for improvement is appreciated 

Refs #

### Added more support for different cases: 
| Case | ref |
| - | - |
| Buy with a credit card, for a custom recipient with valid walelt address | credit-card-with-valid-address |
| Buy with a credit card, for a custom recipient with invalid walelt address | credit-card-with-invalid-address |
| Buy with a credit card, without `advanced mode` | credit-card-basic |
| Buy for custom recipient, if user already have a key | user-with-already-key |
| Buy for custom recipient with invalid wallet address, if user already have a key  | user-with-key-advanced-invalid |
| Buy for custom recipient with valid wallet address, if user already have a key   | user-with-key-advanced-valid |

https://user-images.githubusercontent.com/20865711/157472284-907986e0-0439-449d-bf74-04fc6ff3b260.mov

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/components/interface/checkout/CryptoCheckout.tsx [hunk 1]
```diff
@@ -126,6 +132,30 @@ export const CryptoCheckout = ({
     getBalance()
   }, [account, lock.address, walletNetwork])
 
+  useEffect(() => {
+    if (isAdvanced) return
+    setRecipientValid(true)
+    setRecipient('')
+  }, [isAdvanced])
+
+  const onRecipientChange = useCallback(
```
Comment (searchableguy) [inline unlock-app/src/components/interface/checkout/CryptoCheckout.tsx:141]: any reason for useCallback? This function doesn't have any dependencies.

File: unlock-app/src/components/interface/checkout/CryptoCheckout.tsx [hunk 2]
```diff
@@ -137,13 +167,41 @@ export const CryptoCheckout = ({
...
+          <Label>Recipient</Label>
+          <Input
+            type="text"
+            placeholder="Recipient address"
+            name="recipient"
+            onChange={onRecipientChange}
+            style={{ marginBottom: '0.2rem' }}
+          />
+          {recipient?.length ? (
+            recipientValid ? (
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/CryptoCheckout.tsx:184]: I am not sure we nee to show this message. Let's assume that this is valid unless it is marked as invalid.

Comment (julien51) [inline unlock-app/src/components/interface/checkout/CryptoCheckout.tsx:184]: I am not sure we nee to show this message. Let's assume that this is valid unless it is marked as invalid.

Comment (kalidiagne) [reply unlock-app/src/components/interface/checkout/CryptoCheckout.tsx:188]: oh, let's only show the error message if when is invalid! 👍

**Review by julien51 — APPROVED** [review]
Very clean! Thanks!

**Comment by kalidiagne** [comment]
Just pushed new changes where i covered all other cases like `buy with credit card`, I also updated the PR description with more details and screenshots

**Comment by julien51** [comment]
LGTM! @kalidiagne Please push branches to this one (not to your fork) as the CI job will deploy the site on a "draft" so we can easily test things out :) (for next time!)

**Review by julien51 — APPROVED** [review]

**Pull request merged**